### PR TITLE
Add open/close sounds to context menus

### PR DESCRIPTION
**Pull request by nekodex**
Adds open/close sounds to context menus (using dropdown sounds... for now?).

There's some issues when nested open submenus are closed via either:
- changing top-level item selection
- opening the context menu elsewhere

Doing these when nested submenus are open will result in the close sample playing for each submenu... simultaneously. Not sure how much of that is my fault vs quirks of `Menu` in o!f though, thus I'm making this PR a bit prematurely to figure it out.

partially addresses #14082

### CODE REVIEW

**Comment by frenzibyte** [comment]
> Doing these when nested submenus are open will result in the close sample playing for each submenu... simultaneously. Not sure how much of that is my fault vs quirks of Menu in o!f though, thus I'm making this PR a bit prematurely to figure it out.

The only way I can think of resolving this is doing it as you have it but from framework, i.e. passing a boolean in the `AnimateClose` indicating whether the parent menu is/will also be closed, to not play the close sample in that case.

Though it kinda feels like `AnimateClose` isn't supposed to be invoked on menus below the top closed one, given that menus contain their submenus and fade out with them on animate close. But not sure whether that would be an expected behaviour.

**Comment by bdach** [comment]
> Though it kinda feels like `AnimateClose` isn't supposed to be invoked on menus below the top closed one

Disagree with this. Closing the top-level menu is not the only thing that should trigger an animation. Backing out of a nested menu should also run a fade animation of it.

**Comment by bdach** [comment]
I've come up with an alternative solution to the close samples layering when multiple levels of menu are closed in one frame, but not sure if acceptable. Here's a patch to look at.

```diff
diff --git a/osu.Game/Graphics/Cursor/OsuContextMenuContainer.cs b/osu.Game/Graphics/Cursor/OsuContextMenuContainer.cs
index e64d2d98f4..171ad4ee65 100644
--- a/osu.Game/Graphics/Cursor/OsuContextMenuContainer.cs
+++ b/osu.Game/Graphics/Cursor/OsuContextMenuContainer.cs
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using osu.Framework.Allocation;
 using osu.Framework.Graphics.Cursor;
 using osu.Framework.Graphics.UserInterface;
 using osu.Game.Graphics.UserInterface;
@@ -9,6 +10,14 @@ namespace osu.Game.Graphics.Cursor
 {
     public class OsuContextMenuContainer : ContextMenuContainer
     {
+        [Cached]
+        private OsuContextMenuSamples samples = new OsuContextMenuSamples();
+
+        public OsuContextMenuContainer()
+        {
+            AddInternal(samples);
+        }
+
         protected override Menu CreateMenu() => new OsuContextMenu(true);
     }
 }
diff --git a/osu.Game/Graphics/UserInterface/OsuContextMenu.cs b/osu.Game/Graphics/UserInterface/OsuContextMenu.cs
index 7cf8b3eca8..cf201b18b4 100644
--- a/osu.Game/Graphics/UserInterface/OsuContextMenu.cs
+++ b/osu.Game/Graphics/UserInterface/OsuContextMenu.cs
@@ -4,8 +4,6 @@
 using osuTK.Graphics;
 using osu.Framework.Allocation;
 using osu.Framework.Audio;
-using osu.Framework.Audio.Sample;
-using osu.Framework.Extensions;
 using osu.Framework.Extensions.Color4Extensions;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Effects;
@@ -16,16 +14,15 @@ namespace osu.Game.Graphics.UserInterface
     public class OsuContextMenu : OsuMenu
     {
         private const int fade_duration = 250;
-        private Sample sampleOpen;
-        private Sample sampleClose;
-        private Sample sampleClick;
+
+        [Resolved]
+        private OsuContextMenuSamples samples { get; set; }
 
         // todo: this shouldn't be required after https://github.com/ppy/osu-framework/issues/4519 is fixed.
         private bool wasOpened;
         private readonly bool playClickSample;
-        private readonly Menu parentMenu;
 
-        public OsuContextMenu(bool playClickSample = false, Menu parentMenu = null)
+        public OsuContextMenu(bool playClickSample = false)
             : base(Direction.Vertical)
         {
             MaskingContainer.CornerRadius = 5;
@@ -41,16 +38,12 @@ public OsuContextMenu(bool playClickSample = false, Menu parentMenu = null)
             MaxHeight = 250;
 
             this.playClickSample = playClickSample;
-            this.parentMenu = parentMenu;
         }
 
         [BackgroundDependencyLoader]
         private void load(OsuColour colours, AudioManager audio)
         {
             BackgroundColour = colours.ContextMenuGray;
-            sampleClick = audio.Samples.Get($"UI/{HoverSampleSet.Default.GetDescription()}-select");
-            sampleOpen = audio.Samples.Get(@"UI/dropdown-open");
-            sampleClose = audio.Samples.Get(@"UI/dropdown-close");
         }
 
         protected override void AnimateOpen()
@@ -58,10 +51,10 @@ protected override void AnimateOpen()
             this.FadeIn(fade_duration, Easing.OutQuint);
 
             if (playClickSample)
-                sampleClick?.Play();
+                samples.PlayClickSample();
 
             if (!wasOpened)
-                sampleOpen?.Play();
+                samples.PlayOpenSample();
 
             wasOpened = true;
         }
@@ -70,18 +63,12 @@ protected override void AnimateClose()
         {
             this.FadeOut(fade_duration, Easing.OutQuint);
 
-            if (parentMenu?.State == MenuState.Closed)
-            {
-                wasOpened = false;
-                return;
-            }
-
             if (wasOpened)
-                sampleClose?.Play();
+                samples.PlayCloseSample();
 
             wasOpened = false;
         }
 
-        protected override Menu CreateSubMenu() => new OsuContextMenu(false, this);
+        protected override Menu CreateSubMenu() => new OsuContextMenu();
     }
 }
diff --git a/osu.Game/Graphics/UserInterface/OsuContextMenuSamples.cs b/osu.Game/Graphics/UserInterface/OsuContextMenuSamples.cs
new file mode 100644
index 0000000000..d67ea499e5
--- /dev/null
+++ b/osu.Game/Graphics/UserInterface/OsuContextMenuSamples.cs
@@ -0,0 +1,35 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Audio;
+using osu.Framework.Audio.Sample;
+using osu.Framework.Extensions;
+using osu.Framework.Graphics;
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public class OsuContextMenuSamples : Component
+    {
+        private Sample sampleClick;
+        private Sample sampleOpen;
+        private Sample sampleClose;
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours, AudioManager audio)
+        {
+            sampleClick = audio.Samples.Get($@"UI/{HoverSampleSet.Default.GetDescription()}-select");
+            sampleOpen = audio.Samples.Get(@"UI/dropdown-open");
+            sampleClose = audio.Samples.Get(@"UI/dropdown-close");
+        }
+
+        public void PlayClickSample() => Scheduler.AddOnce(playClickSample);
+        private void playClickSample() => sampleClick.Play();
+
+        public void PlayOpenSample() => Scheduler.AddOnce(playOpenSample);
+        private void playOpenSample() => sampleOpen.Play();
+
+        public void PlayCloseSample() => Scheduler.AddOnce(playCloseSample);
+        private void playCloseSample() => sampleClose.Play();
+    }
+}
```

tl;dr: sample playback moved to component cached at the menu container level, with scheduling logic inside to play at most once per frame. Eliminates the parent menu tracking, which is nice; the `playClickSample` has to remain for now, though. (Framework has `TopLevelMenu` but that is actually something else yet.)

**Comment by frenzibyte** [comment]
> > Though it kinda feels like `AnimateClose` isn't supposed to be invoked on menus below the top closed one
> 
> Disagree with this. Closing the top-level menu is not the only thing that should trigger an animation. Backing out of a nested menu should also run a fade animation of it.

By "menus below the top closed one", I mean going more than one menu up instantaneously:

https://user-images.githubusercontent.com/22781491/128605147-1d6b67c7-5ce6-4040-a85d-51223b2b38a0.mp4

But your alternative suggestion of using `AddOnce` scheduling works as well, could even go for configuring `PlaybackConcurrency` instead if that's feasible, but scheduling works for now.

**Comment by smoogipoo** [comment]
I like @bdach 's suggestion. A bit of a shame that it's required in the first place so this definitely needs to be a consideration for future work on menus.

**Comment by bdach** [comment]
I've pushed in the aforementioned patch here.

**Review by frenzibyte — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

**Review by peppy — APPROVED** [review]
